File: engine/src/main/java/com/terralite/engine/world/World.java
package com.terralite.engine.world;

import com.terralite.engine.chunk.Chunk;
import com.terralite.engine.chunk.ChunkPos;
import com.terralite.engine.chunk.ChunkStorage;
import com.terralite.engine.chunk.InMemoryChunkStorage;
import com.terralite.engine.entity.EntityManager;
import com.terralite.engine.entity.EntitySpawner;
import com.terralite.engine.entity.InMemoryEntityManager;
import com.terralite.engine.physics.Aabb;
import com.terralite.engine.terrain.BlockPos;
import com.terralite.engine.terrain.BlockState;
import com.terralite.engine.terrain.BlockStorage;
import com.terralite.engine.terrain.MultiblockBlockStorage;
import com.terralite.engine.terrain.SparseBlockStorage;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public final class World implements WorldAccess {
    private final ChunkStorage chunks;
    private final EntityManager entities;
    private final BlockStorage blocks;

    public World() {
        this(new InMemoryChunkStorage(), new InMemoryEntityManager(), new SparseBlockStorage());
    }

    public World(BlockStorage blocks) {
        this(new InMemoryChunkStorage(), new InMemoryEntityManager(), blocks);
    }

    public World(ChunkStorage chunks) {
        this(chunks, new InMemoryEntityManager(), new SparseBlockStorage());
    }

    public World(ChunkStorage chunks, EntityManager entities) {
        this(chunks, entities, new SparseBlockStorage());
    }

    public World(ChunkStorage chunks, EntityManager entities, BlockStorage blocks) {
        this.chunks = Objects.requireNonNull(chunks, "chunks");
        this.entities = Objects.requireNonNull(entities, "entities");
        this.blocks = Objects.requireNonNull(blocks, "blocks");
    }

    public ChunkStorage chunks() {
        return chunks;
    }

    public EntityManager entities() {
        return entities;
    }

    public BlockStorage blocks() {
        return blocks;
    }

    public EntitySpawner spawner() {
        return new EntitySpawner(this);
    }

    public BlockState getBlock(BlockPos pos) {
        return blocks.get(pos);
    }

    public void setBlock(BlockPos pos, BlockState state) {
        blocks.set(pos, state);
    }

    public BlockState removeBlock(BlockPos pos) {
        return blocks.remove(pos);
    }

    public Collection<BlockPos> collisionBlockPositions() {
        if (blocks instanceof MultiblockBlockStorage multiblockBlocks) {
            List<BlockPos> positions = new ArrayList<>();
            for (BlockPos origin : multiblockBlocks.positions()) {
                positions.addAll(multiblockBlocks.occupiedPositions(origin));
            }
            return List.copyOf(positions);
        }
        return blocks.positions();
    }

    public Collection<Aabb> blockCollisionBoxes() {
        return collisionBlockPositions().stream()
                .map(World::blockBounds)
                .toList();
    }

    public Chunk putChunk(Chunk chunk) {
        return chunks.put(chunk);
    }

    public Chunk removeChunk(ChunkPos pos) {
        return chunks.remove(pos);
    }

    @Override
    public Optional<Chunk> getChunk(ChunkPos pos) {
        return chunks.get(pos);
    }

    @Override
    public Chunk requireChunk(ChunkPos pos) {
        return chunks.require(pos);
    }

    @Override
    public boolean containsChunk(ChunkPos pos) {
        return chunks.contains(pos);
    }

    @Override
    public Collection<ChunkPos> chunkPositions() {
        return chunks.positions();
    }

    private static Aabb blockBounds(BlockPos pos) {
        return new Aabb(pos.x(), pos.y(), pos.z(), pos.x() + 1.0, pos.y() + 1.0, pos.z() + 1.0);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/world/World.kt
package com.terralite.engine.world

import com.terralite.engine.chunk.Chunk
import com.terralite.engine.chunk.ChunkPos
import com.terralite.engine.chunk.ChunkStorage
import com.terralite.engine.chunk.InMemoryChunkStorage
import com.terralite.engine.entity.EntityManager
import com.terralite.engine.entity.EntitySpawner
import com.terralite.engine.entity.InMemoryEntityManager
import com.terralite.engine.physics.Aabb
import com.terralite.engine.terrain.BlockPos
import com.terralite.engine.terrain.BlockState
import com.terralite.engine.terrain.BlockStorage
import com.terralite.engine.terrain.MultiblockBlockStorage
import com.terralite.engine.terrain.SparseBlockStorage

class World(
    private val chunks: ChunkStorage,
    private val entities: EntityManager,
    private val blocks: BlockStorage
) : WorldAccess {
    constructor() : this(InMemoryChunkStorage(), InMemoryEntityManager(), SparseBlockStorage())
    constructor(blocks: BlockStorage) : this(InMemoryChunkStorage(), InMemoryEntityManager(), blocks)
    constructor(chunks: ChunkStorage) : this(chunks, InMemoryEntityManager(), SparseBlockStorage())
    constructor(chunks: ChunkStorage, entities: EntityManager) : this(chunks, entities, SparseBlockStorage())

    fun chunks(): ChunkStorage = chunks
    fun entities(): EntityManager = entities
    fun blocks(): BlockStorage = blocks
    fun spawner(): EntitySpawner = EntitySpawner(this)

    fun getBlock(pos: BlockPos): BlockState = blocks.get(pos)
    fun setBlock(pos: BlockPos, state: BlockState) = blocks.set(pos, state)
    fun removeBlock(pos: BlockPos): BlockState = blocks.remove(pos)

    fun collisionBlockPositions(): Collection<BlockPos> {
        if (blocks is MultiblockBlockStorage) {
            val positions = mutableListOf<BlockPos>()
            for (origin in blocks.positions()) {
                positions += blocks.occupiedPositions(origin)
            }
            return positions.toList()
        }
        return blocks.positions()
    }

    fun blockCollisionBoxes(): Collection<Aabb> =
        collisionBlockPositions().map { pos ->
            Aabb(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble(),
                 pos.x + 1.0, pos.y + 1.0, pos.z + 1.0)
        }

    fun putChunk(chunk: Chunk): Chunk = chunks.put(chunk)
    fun removeChunk(pos: ChunkPos): Chunk = chunks.remove(pos)

    override fun getChunk(pos: ChunkPos): Chunk? = chunks.get(pos)
    override fun requireChunk(pos: ChunkPos): Chunk = chunks.require(pos)
    override fun containsChunk(pos: ChunkPos): Boolean = chunks.contains(pos)
    override fun chunkPositions(): Collection<ChunkPos> = chunks.positions()
}
